cmd/aws-vpc-cni-init/main.go

```diff
@@ -0,0 +1,177 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// The aws-node initialization
+package main
+
+import (
+	"os"
+
+	"github.com/aws/amazon-vpc-cni-k8s/utils/cp"
+	"github.com/aws/amazon-vpc-cni-k8s/utils/imds"
+	"github.com/aws/amazon-vpc-cni-k8s/utils/sysctl"
+	log "github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	defaultHostCNIBinPath = "/host/opt/cni/bin"
+	vpcCniInitDonePath    = "/vpc-cni-init/done"
+	metadataLocalIP       = "local-ipv4"
+	metadataMAC           = "mac"
+
+	envDisableIPv4TcpEarlyDemux = "DISABLE_TCP_EARLY_DEMUX"
+	envEnableIPv6               = "ENABLE_IPv6"
+	envHostCniBinPath           = "HOST_CNI_BIN_PATH"
+)
+
+func getEnv(env, def string) string {
+	if val, ok := os.LookupEnv(env); ok {
+		return val
+	}
+	return def
+}
+
+func main() {
+	os.Exit(_main())
+}
+
+func _main() int {
+	log.Debug("Started Initialization")
+	pluginBins := []string{"loopback", "portmap", "bandwidth", "aws-cni-support.sh"}
+	var err error
+	for _, plugin := range pluginBins {
+		if _, err = os.Stat(plugin); err != nil {
+			log.WithError(err).Fatalf("Required executable: %s not found\n", plugin)
+			return 1
+		}
+	}
+
+	hostCNIBinPath := getEnv(envHostCniBinPath, defaultHostCNIBinPath)
+
+	log.Infof("Copying CNI plugin binaries ...")
+	err = cp.InstallBinaries(pluginBins, hostCNIBinPath)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to install binaries")
+		return 1
+	}
+
+	log.Infof("Copied all CNI plugin binaries to %s\n", hostCNIBinPath)
+
+	var primaryMAC string
+	primaryMAC, err = imds.GetMetaData("mac")
+	if err != nil {
+		log.WithError(err).Fatalf("aws-vpc-cni init failed\n")
+		return 1
+	}
+
+	log.Infof("Found primaryMAC %s", primaryMAC)
+
+	links, err := netlink.LinkList()
+	if err != nil {
+		log.WithError(err).Fatalf("Failed to list links\n")
+		return 1
+	}
+
+	var primaryIF string
+	for _, link := range links {
+		if link.Attrs().HardwareAddr.String() == primaryMAC {
+			primaryIF = link.Attrs().Name
+			break
+		}
+	}
+
+	if primaryIF == "" {
+		log.Errorf("Failed to retrieve primary IF")
+		return 1
+	}
+	log.Infof("Found primaryIF %s", primaryIF)
+
+	sys := sysctl.New()
+	// Configure rp_filter in loose mode
+	entry := "net/ipv4/conf/" + primaryIF + "/rp_filter"
+	err = sys.SetSysctl(entry, 2)
+	if err != nil {
+		log.WithError(err).Fatalf("Failed to set rp_filter for %s\n", primaryIF)
+		return 1
+	}
+
+	val, _ := sys.GetSysctl(entry)
+	log.Infof("Updated %s to %d", entry, val)
+
+	// Enable or disable TCP early demux based on environment variable
+	disableIPv4EarlyDemux := getEnv(envDisableIPv4TcpEarlyDemux, "false")
+	entry = "net/ipv4/tcp_early_demux"
+	if disableIPv4EarlyDemux == "true" {
+		err = sys.SetSysctl(entry, 0)
+		if err != nil {
+			log.WithError(err).Fatalf("Failed to disable tcp_early_demux\n")
+			return 1
+		}
+	} else {
+		err = sys.SetSysctl(entry, 1)
+		if err != nil {
```

Comment:
the original shell-based script have '-e'(Use this option to ignore errors about unknown keys), while Golang version don't. wondering why we have it in the shell version

Reply:
Ah nice catch, looks like this comes from https://github.com/aws/amazon-vpc-cni-k8s/issues/1241 . Not sure if we still need to support kernels this old anymore, but it seems worth catching this error